### PR TITLE
remove spurious usages of virtual keyword

### PR DESCRIPTION
**Pull request by alexis-pereda**
This PR removes several usages of the `virtual` keyword.

For those who would want to know why:
[C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-override)

By default, if `B` inherits `A` and both provide a member function `f`, calling it through a reference of type `A` or a pointer of type `A*` which actually refers to a `B` (or holds the address of a `B` for the pointer case) will cause the version from class `A` to be used.
Adding `virtual` changes the behaviour to make it so it calls the member function of class `B`.

Making `f` virtual in `A` makes it `virtual` in all inheriting class, so it is not necessary to specify the `virtual` keyword except in class `A`.
However, to indicate in subclasses that a function is `virtual` to the human, before C++11 it was only possible to either add a comment or put the `virtual` keyword anyway.
Since C++11, a new keyword exists: `override`.
It can only be used on member functions that are `virtual`, so the compiler will check that for us (it helps avoiding spelling/type/constness mistakes).

Putting both `virtual` and `override` on the same member function declaration breaks the effect of `override` because even if the function does not exist or is not virtual in the parent class, it will compile as it is marked virtual.

TL; DR: any member function must be marked either `virtual` or `override` (or none), never both.

### CODE REVIEW

**Comment by dsarrut** [comment]
It is tagged as "draft", should we merge or wait?

**Comment by alexis-pereda** [comment]
I was waiting for all the tests to finish to be sure. it can be merged.
